OPHKIOS-184: asiakastiedot listat frontend fixes

saku-koodari opened the pull request:
## Yhteenveto

Sovitettu frontti ja bäkkäri yhteen. Lisätty testidataa. Ei ole täysin samanlaista kuin spekseissä, testidatasta puuttuu jotain päivämääriä. Se pitäisi olla silti testaamiseen ihan ok, koska MSW kattaa nekin tapaukset

## Lisätiedot

_Tehdyt muutokset..._

## Huomautukset

_Muut mahdolliset huomautukset..._

## Check-lista

- [ ] yarn.lock päivitetty
- [ ] Käännösten Excel-tiedosto päivitetty
- [ ] UI-muutoksia testattu eri selaimilla ja mobiilinäkymässä

saku-koodari committed:
YKI(Frontend): exam supports now different languages

saku-koodari committed:
YKI(Frontend): Ignore build files of yki yarn:build

saku-koodari committed:
YKI(Frontend): Merge registration types in response

saku-koodari committed:
wip: YKI(Frontend): refactor frontend serialization to map backend data correctly

saku-koodari committed:
YKI(Frontend): Rename schoolName to name

saku-koodari committed:
YKI(Backend): Get liftedFromQueueAt - date time from backend

saku-koodari committed:
YKI(Backend): Add Registration.Kind to Response

saku-koodari committed:
YKI(Backend): Add Registration.expiresAt to Model

saku-koodari committed:
YKI(Frontend): Update interface to match with Backend DTO

saku-koodari committed:
YKI(Frontend): Fix MSF fixure customerDetails

saku-koodari committed:
YKI(Frontend): Refactor customerDetails of serialization

saku-koodari committed:
YKI(Frontend): Logic for QueueSpotOffered

saku-koodari committed:
YKI(Frontend): PastExam - RegistrationState

saku-koodari committed:
YKI(Frontend): Disable dot rule on non-clerk

saku-koodari committed:
YKI(Frontend): Use SerializationUtils instead of this

saku-koodari committed:
YKI(Frontend): Move upcoming registrations to the future

so that during our tests, we can see the data as expected the next 10 years

saku-koodari committed:
YKI(Frontend): Remove FREE_REGISTRATION_PENDING - references

saku-koodari committed:
YKI(Frontend): Make the RegistrationStatus - type more relaxed

saku-koodari committed:
YKI(Frontend): rename dueDate -> expiresAt

saku-koodari committed:
YKI(Frontend): Refactor QueueSpotOffered - logic

saku-koodari committed:
YKI(Backend): Flatten registration status

saku-koodari committed:
YKI(Backend): Allow registrationDate to be optional

because our database allows optional values.

saku-koodari committed:
YKI(Frontend): Flatten registration status

saku-koodari committed:
YKI(Frontend): Take nullable examination date into account

saku-koodari committed:
YKI(Backend): Check data from local repository first

This is show proper error message:
 - If the person is missing from our db, it should be 404
 - if the person is in our db, but missing from onr, it is some kind of data issue

saku-koodari committed:
YKI(Backend): Update RegistrationState.java

saku-koodari committed:
YKI(Backend): Refactor code

saku-koodari committed:
YKI(Backend): Fix exam level

saku-koodari committed:
YKI(Backend): Update test data

saku-koodari committed:
YKI(Frontend): Fix lint errors

saku-koodari committed:
YKI(Frontend): Update the date to match with the mock data

saku-koodari committed:
YKI(Frontend): Update fixed clock

saku-koodari commented on frontend/packages/yki/src/tests/cypress/support/index.ts:
```diff
@@ -16,7 +16,7 @@ Cypress.on('test:after:run:async', async () => {
 beforeEach(() => {
   // Use fixed date for tests
-  const fixedDateForTests = dayjs('2022-09-27T16:00:00+0200');
+  const fixedDateForTests = dayjs('2025-09-27T16:00:00+0200');
```
Päivitin testien kellon.  Koska vertaan päivämääriä dayjs.now:n [perusteella](https://github.com/Opetushallitus/kieli-ja-kaantajatutkinnot/pull/925/files#diff-a857e55379ed6b891ddcbc56156c1c47e5927a1ab4a43f885e8bef794f4d18bcR460). 

Ongelma tulee siis siitä, kun kello oli asetettu täällä vuoteen 2022, ja savutestatessa kello on luonnollisesti nyt eli 2025. Toinen osa mokkidatasta on tyyliin vuonna 2025 ja se on tulkittu olevan menneisyydessä. 

Siis toinen vaihtoehto korjata nämä olisi kovakoodata menneisyydessä olevat datat esim vuoteeen 2022

saku-koodari replied:
Nyt kun mietin asiaa, se voisi olla ehkä järkevämpi vaihtoehnto, vaikka tämäkään muutos ei hajottanut mitään.

saku-koodari committed:
YKI(Backend): Add missing expires_at - column to new backend

pkoivisto commented on frontend/packages/yki/src/interfaces/clerkCustomer.ts:
```diff
@@ -13,41 +19,42 @@ export interface CustomerPerson {
+  lang: AppLanguage;
+};
+
+export type Exam = {
```
Tämä `Exam`-tyyppi viitannee oikeasti ilmoittautumiseen, jos tuolla kerran on `registrationStatus`- ja `registrationDate`-kentät? Nimeämistä voisi ehkä olla hyvä vielä pohtia niin, että se olisi ymmärrettävämpi.

Mikäli virkailija- ja julkipuolen tyypit tai konseptit clashaavat , niin yksi vaihtoehto voisi olla harkita vaikkapa eri näkymissä tarvittujen tyyppien prefiksausta tai jotain muuta nimeämiskonventiota; esim. VKT:ssa tätä on jossain määrin hanskattu `Clerk`- ja `Public`-etuliitteillä tyyppien nimissä.

pkoivisto commented on frontend/packages/yki/src/utils/serialization.ts:
```diff
@@ -437,40 +457,159 @@ export class SerializationUtils {
+      // examinationDate is calculated by fetching the date from the latest examPayment. A date from freeRegistration works as the fallback value, if there is no value. Technically freeRegistration may be also null.
+      // Practically, if the examinationDate is missing,
+      // it's data issue and then probably the data should be fixed.
+      examinationDate: dayjs(registration.examinationDate),
```
`examinationDate` kuulostaa konseptina tutkinnon päivämäärältä, joka tietysti on eri asia kuin ilmoittautumisen ajankohta -> nimeä voisi muuttaa. Itse tutkinnon päivähän taas löytyy `exam_date`-taulusta.

saku-koodari replied:
Katsoin koodia, ja tuo mapataan frontissa tutkinnnon -pvm - kenttään. Teen korjaukset tähän.

saku-koodari replied:
mulla oli käyny ajatuspieru tuota kommenttia kirjoittaessa. Kävin koodin läpi, ja oon tosiaa varmaan ajatellut registrationDate - kenttää tätä kirjoittaessa. Nimesin tuon kentän joka tapauksessa uudelleen -> examDate ja poistin tuon kommntin. 

Toi kommentti tosiaan liittyi registrationDate - kenttään. Se kenttä on nyt koodissa optional. Mun mielestä siinä kentässä ei tarvii frontin näkökulmasta tarvii välttämättä tietää muuta bäkkäristä kuin se, että sieltä voi tulla arvo tai olla tulematta arvo

kimmotaskinen-knowit commented on frontend/packages/yki/src/components/clerkCustomer/CustomerExamListings.tsx:
```diff
@@ -198,7 +216,7 @@ export const CustomerExamListings = ({
               {t('values.queueSpotOffered.dueDate', {
                 dueDate: DateUtils.formatOptionalDate(
-                  queueSpotOffered.dueDate,
+                  queueSpotOffered.expiresAt,
```
käännösavaimen voisi päivittää mätsäämään myös tota poimittua propsia

saku-koodari committed:
YKI(Frontend): Rename dueDate -> expiresAt, also on resource keys

saku-koodari committed:
YKI(Frontend/Backend): Rename examinationDate -> examDate

saku-koodari committed:
YKI(Frontend): Rename Exams -> Registration

In the context of clerk customer, to make clarify better of the symbols.
Refactored names
 -> Exam -> AdmissionedRegistration
 -> QueuedExam -> QueuedRegistration
 -> PastExam -> PastRegistration

saku-koodari committed:
YKI(Frontend): Show not queued (upcoming) registrations as admissioned registratons

The data model theoretically supports other registration.kind - values than Admissioned and Queued.
However new registrations will have either .kind === Admissioned|Queued

This change will show the legacy types as Admissioned

pkoivisto commented on frontend/packages/yki/src/tests/msw/fixtures/customerDetails.ts:
```diff
@@ -66,7 +66,7 @@ const helsinki = [
 const defaultRegistrations: RegistrationResponse[] = [
   {
-    examinationDate: '2035-09-01T00:00:00.000Z',
+    examDate: '2035-09-01T00:00:00.000Z',
```
Oliko tää tarkoituksella eri muotoa kuin muualla mock-datassa? Vai pitäisikö mieluummin olla ihan vain päivämäärä `2035-09-01`?

pkoivisto commented on frontend/packages/yki/src/tests/msw/fixtures/customerDetails.ts:
```diff
@@ -91,7 +91,7 @@ const defaultRegistrations: RegistrationResponse[] = [
   },
   {
-    examinationDate: '2035-011-30',
+    examDate: '2035-011-30',
```
Pitäisi varmaan olla joko `2035-01-30` tai `2035-11-30`?

pkoivisto commented on backend/yki/src/main/java/fi/oph/yki/api/dto/clerk/ClerkCustomerRegistrationDTO.java:
```diff
@@ -1,15 +1,24 @@
+  Optional<LocalDate> examPaymentPaidAt,
+  Optional<LocalDate> registrationDate,
+  RegistrationKind kind,
+  LocalDateTime liftedFromQueueAt,
```
Olisikohan järkevää, että nää ehkä mahdollisesti löytyvät kentät - vissiin siis liftedFromQueueAt, examPaymentPaidAt, registrationDate - olisi joko kaikki konsistentisti Optional tai ei yksikään?

pkoivisto commented on frontend/packages/yki/src/tests/msw/fixtures/customerDetails.ts:
```diff
@@ -1,192 +1,207 @@
+    registrationState: RegistrationStates.Submitted,
+    registrationDate: '2035-05-06',
+    kind: RegistrationKind.Queue,
+    liftedFromQueueAt: '2035-05-06', // kentässä on arvo => jonopaikkaa tarjottu
```
Tässä on hyvä huomata, että jos jonopaikkaa on tarjottu, ilmoittautuminen siirtyy aina jonopaikasta varsinaiseksi ilmoittautumiseksi -> kind: `RegistrationKind.Admission`. Tässä oleva kombinaatio, jossa `kind = 'QUEUE'` ja `liftedFromQueueAt != null` on siis mahdoton.
 
Jos ilmoittautuminen on maksuton, ilmoittautuminen muuttuu tällöin suoraan vahvistetuksi (state = 'COMPLETED'); jos maksu vaaditaan niin ilmoittautuminen jää vielä odottamaan maksua (state = 'SUBMITTED').

saku-koodari committed:
YKI(Frontend): Fix typos from mock data

saku-koodari committed:
YKI(Frontend): Revert clock and set past dates more back in time

saku-koodari committed:
YKI(Backend): All nullable fields as Optional and date -fields as LocalDate - type

pkoivisto reviewed (approved):
Ok, kunhan testit menee läpi.

saku-koodari committed:
YKI(Frontend): Add TODO-comment

saku-koodari committed:
YKI(Frontend): Fix tests
